tree-sitter-tips.md

```diff
@@ -0,0 +1,79 @@
+# Tree-sitter tips and tricks
+
+Language development using [Tree-Sitter](https://github.com/tree-sitter/tree-sitter) can be an arduous task. These tips are intended to help bring empiricism to the process by reducing trial and error (or at least introducing some guard rails).
+
+### :bulb: Rely on Language Specs, but not too much.
+There are language specifications (like [this one for Java](https://docs.oracle.com/javase/specs/jls/se9/html/index.html)) that comprehensively describe a language. Specs provide the lexical and syntactic structure of a language in addition to its components (ex., types, variables, declarations, etc.) and how they interconnect. This can serve as a great baseline for writing your grammar. However, mirroring the spec 100% is not always the best approach, especially when its structure does not align with the parse tree you'd like to generate using tree-sitter. In this situation, it's better to deviate from the spec and prioritize generating a good parse tree. One way to do this may be to find an LR parser for the language you're working on ([such as this one for Java](https://github.com/javaparser/javaparser)) to see how they've modeled their tree outputs. Studying tree outputs from other LR parsers can be a good way to guide your own grammar development. 
+
+### :bulb: What does a "good" parse tree look like?
+
+Here are some attributes of a "pretty" tree:
+
+:white_check_mark: Not overly verbose.
+
+:white_check_mark: Everything is outlined clearly.
+
+:white_check_mark: Communicates good information needed further along in the pipeline.
+
+![image](https://user-images.githubusercontent.com/875834/34972577-41bc12ac-fa37-11e7-9a12-a05d18b6f6e4.png)
+
+_Example of a good tree (WIP -- better examples welcome!):_
```

Comment:
Might be worth looking at the corpus of examples in something like tree-sitter-javascript for this.